### ISSUE TITLE
Up-to-date check - <FastUpToDateCheckIgnoresKinds> doesn't work?

### ISSUE THREAD

**Issue by pascal-910** [dotnet/project-system#8935]
## Visual Studio Version
17.5.3 Professional

## Summary
This problem is related to "**Filtering inputs and outputs**" described in [up-do-dat-check.md](https://github.com/dotnet/project-system/blob/dev17.3.x/docs/up-to-date-check.md).

Maybe I miss somthing but as I understand I should be able to ignore/exclude changes to specified input files that will otherwise lead update-to-check to set them to be outdated and thus force a build.

## Steps to Reproduce

1. Set Logging Level to "Verbose" in Options - Projects and Solutions - Up to Date Checks
3. Create a .NET6 console application or class library
4. Add a new class Foo.cs
5. Add the following to .csproj
```
<ItemGroup>
	<UpToDateCheckInput Include="Foo.cs" Kind="XXX" />
</ItemGroup>
<PropertyGroup>
	<FastUpToDateCheckIgnoresKinds>XXX</FastUpToDateCheckIgnoresKinds>
</PropertyGroup>
```
6. Build two times. Output says everything is up-to-date:

========== Build: 0 succeeded, 0 failed, 1 up-to-date, 0 skipped ==========

4. Change anything in Foo.cs
7. Build

## Expected Behavior
Project should be up-to-date and not built again.

## Actual Behavior
Output says Foo.cs is not up-to-date:

```
1>FastUpToDate: Input Compile item 'D:\Entw\ConsoleApp1\ConsoleApp1\Foo.cs' is newer (2023-03-24 11:05:08.954) than earliest output 'D:\Entw\ConsoleApp1\ConsoleApp1\obj\Debug\net6.0\ConsoleApp1.pdb' (2023-03-24 11:00:36.970), not up-to-date. (ConsoleApp1)
...
========== Build: 1 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
```

**Comment by drewnoakes** [comment]
Notice in the log message that it's mentioning how `Foo.cs` is a `Compile` item?

Your example should work of you replace the addition of an `UpToDateCheckInput` with an update of that `Compile` item.

I'm curious to learn more about the scenario you're using this feature for, if you are happy to share.

**Comment by pascal-910** [comment]
> Yes, that's the line. It mentioned it's a "Compile" input. These items are always considered by the up-to-date check and don't need corresponding UpToDateCheckInput items.

Ok, thanks, now know what you mean  (difference between automatically "Compile" inputs and manually added "UpToDateCheckInput" items).

> Does this work for you?

Unfortunately no - I also changed my example to have a `Foo.txt` defined as a `UpToDateCheckInput`  instead of a `Foo.cs` to see if this works leastwise (see below):
BUT it seems that putting a `Kind="XXX"` attribute to either a "Copile" or a manually added "UpToDateCheckInput" **takes absolutely no effect?!**


See this simplified example:
## Steps to Reproduce
1) Create Console application + Foo.txt

```
<ItemGroup>
<UpToDateCheckInput Include="Foo.txt" Kind="XXX" />
</ItemGroup>

<PropertyGroup>
<FastUpToDateCheckIgnoresKinds>XXX</FastUpToDateCheckIgnoresKinds>
</PropertyGroup>
```
2) Build
3) Modify Foo.txt
4) Build several times

## Actual Behavior
Project is always built if Foo.txt is newer than date of the last build.
```
1>FastUpToDate: Input UpToDateCheckInput item 'D:\Test\Foo.txt' is newer (2023-04-06 08:20:20.558) than earliest output 'D:\Test\obj\Debug\net6.0\AAA_Incremental.pdb' (2023-04-06 08:06:09.498), not up-to-date. (Test)
========== Build: 1 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
```

## Expected Behavior
Should consider as `up-to-date` (ignore changes to Foo.txt) as described in the [docs](https://github.com/dotnet/project-system/blob/main/docs/up-to-date-check.md):
> It may be desirable ... to schedule a build for which only a subset of the up-to-date check inputs ... should be considered.

**Comment by drewnoakes** [comment]
I do think we can improve the documentation here as it's missing some explanation around the correct usage of `FastUpToDateCheckIgnoresKinds`. That said, I don't think it's the feature you're looking for.

**Comment by pascal-910** [comment]
> I'm unclear on why you want to add Foo.txt as an UpToDateCheckInput item, only to permanently have it be ignored.

I agree - but this was only a simple example to check whether `FastUpToDateCheckIgnoresKinds` does anything (I couldn't archive by now).

To make things clear: What I understand this feature to be for in my words (maybe I'm totally wrong):

"To configure an input file so that the up-to-date check doesn't look for it on incremental builds - With the result corresponding projects won't be rebuilt if this specified file was modified. This input file may be either a copile item or a manually added input file added with <UpToDateCheckInput>."
This is the purpose isn't it?

According to improved docs I need to pass the `FastUpToDateCheckIgnoresKinds`  as a **global property** (not .csproj) - maybe this is the reason I didn't see any changes...?
-> I'm sorry I stuck but how do I provide this **global property**? I can't call `msbuild.exe p:FastUpToDateCheckIgnoresKinds=xxx` because this way the up-to-date check won't be included, isn't it?!

**Comment by drewnoakes** [comment]
That's a fair summary. Honestly the end user isn't generally going to find this feature helpful. It's more for VS extensions.

From the updated docs:

> For an example, see `SkipAnalyzersGlobalPropertiesProvider` in this repo.

That component allows some builds to skip running analysers during build for increased performance.

**Comment by pascal-910** [comment]
Ok, I see - thanks for the clarification.
Furtunately I've now found another solution for my requirements (code generation during build) using `UpToDateCheckInput` and `UpToDateCheckOutput` together wirh the `Set` attribute and taking these two as Input and Output for my 'code genration' Target.

**Comment by makolyte** [comment]
> I also found an [article ](https://makolyte.com/how-to-make-the-build-up-to-date-check-ignore-your-text-template-generated-output/) describing this feature as the solution to a very similar requirement, but either I miss something or it simply doesn't work :-)

I'm the author of that article. I wrote that last year (Jan '22). The purpose was to use FastUpToDateCheckIgnoresKinds with UpToDateCheckInput to exclude a T4 generated source file from the Up-to-Date check. But during a periodic review recently (June '23), I realized this solution no longer worked. 

Perhaps FastUpToDateCheckIgnoresKinds was not intended to be used for this purpose and the functionality changed in the last 1.5 years. Regardless of why it no longer works, I'm simply going to delete the article (410-Gone). Feel free to remove the now-dead link.
